mozphab/jujutsu.py

```diff
@@ -0,0 +1,471 @@
+            self.__git_repo = Git(self.git_path)
+        else:
+            # TODO: Do we actually only support co-located Git repos now? 🤔
+            raise Error(
+                "Underlying Git repository for Jujutsu could not be found. "
+                "Make sure you are using a colocated Git repository and not "
+                "the pure Jujutsu backend."
+            )
+
+        # Populate common fields expected from a `Repository`
+
+        dot_path = os.path.join(path, ".jj")
+        if not os.path.exists(dot_path):
+            raise ValueError("%s: not a Jujutsu repository" % path)
+        logger.debug("found Jujutsu repo in %s", path)
+        super().__init__(path, dot_path)
+
+        self.vcs = "jj"
+
+        version_output = check_output(["jj", "--version"], split=False)
+        m = re.search(r"jj (\d+\.\d+\.\d+.*)", version_output)
+        if not m:
+            raise Error("Failed to determine Jujutsu version.")
+        self.vcs_version = m.group(0)
+
+        self.revset = None
+        self.branch = None
+
+        # Populate fields unique to this implementation
+        # TODO?
+
+        self.__username = check_output(
+            ["jj", "config", "get", "user.name"], split=False
+        ).rstrip()
+        self.__email = check_output(
+            ["jj", "config", "get", "user.email"], split=False
+        ).rstrip()
+
+    def check_vcs(self) -> bool:
+        if self.args.force_vcs:
+            return True
+
+        if self.is_cinnabar_required and not self.is_cinnabar_installed:
+            raise Error(
+                "Git Cinnabar extension is required to work on this repository."
+            )
+
+        return True
+
+    def before_submit(self):
+        self.__validate_email()
+
+    def set_args(self, args: argparse.Namespace):
+        """Store moz-phab command line args and set the revset."""
+        super().set_args(args)
+
+        # TODO: Did we accidentally change the behavior of `--single` with no `start_rev` specified?
+        is_single = hasattr(self.args, "single") and self.args.single
+
+        # TODO: What if we try to submit something that's immutable?
+        start_rev = None
+        if (
+            hasattr(self.args, "start_rev")
+            and self.args.start_rev != environment.DEFAULT_START_REV
+        ):
+            start_rev = self.args.start_rev
+        else:
+            start_rev = (
+                self.__get_last_mutable_parent()
+                if is_single
+                else self.__get_first_mutable_parent()
+            )
+        if not start_rev:
+            return None
+
+        end_rev = None
+        if is_single:
+            end_rev = start_rev
+        elif (
+            hasattr(self.args, "end_rev")
+            and self.args.end_rev != environment.DEFAULT_END_REV
+        ):
+            end_rev = self.args.end_rev
+        else:
+            end_rev = self.__get_last_mutable_parent()
+
+        self.revset = (start_rev, end_rev)
```

Comment:
suggestion: @ahal and @Mossop raised in internal Slack that it might be a good idea to change this auto-detection to be include the working-copy commit if it is "ready" (i.e., not empty and with a non-empty description).

Reply:
We might also want to respect [`git.private-commits`](https://martinvonz.github.io/jj/latest/config/#set-of-private-commits) here.